staging/src/k8s.io/apimachinery/pkg/util/sets/set.go

```diff
@@ -181,20 +177,52 @@ func (s1 Set[T]) Equal(s2 Set[T]) bool {
 	return len(s1) == len(s2) && s1.IsSuperset(s2)
 }
 
-type sortableSlice[T SetTypes] []T
+type sortableSlice[T SetTypes] struct {
+	slice []T
+	OrderedFuncs[T]
+}
+
+func (s sortableSlice[T]) Len() int {
+	if s.LenFunc == nil {
+		return len(s.slice)
+	}
+	return s.LenFunc(s.slice)
+}
+func (s sortableSlice[T]) Less(i, j int) bool {
+	if s.LessFunc == nil {
+		// unsorted if no less func specified
+		return true
+	}
+	return s.LessFunc(s.slice, i, j)
+}
+func (s sortableSlice[T]) Swap(i, j int) {
+	if s.SwapFunc == nil {
+		s.slice[i], s.slice[j] = s.slice[j], s.slice[i]
+		return
+	}
+	s.SwapFunc(s.slice, i, j)
+}
 
-func (s sortableSlice[T]) Len() int           { return len(s) }
-func (s sortableSlice[T]) Less(i, j int) bool { return lessSetType[T](s[i], s[j]) }
-func (s sortableSlice[T]) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
+// OrderedFuncs allows custom sorting methods like sort.Interface.
+// List will be unsorted if LessFunc absent, in other word, UnsortedList
+// should be used in this scenario.
+type OrderedFuncs[T SetTypes] struct {
+	LenFunc  func(s []T) int
+	LessFunc func(s []T, i, j int) bool
+	SwapFunc func(s []T, i, j int)
+}
 
-// List returns the contents as a sorted T slice.
-func (s Set[T]) List() []T {
-	res := make(sortableSlice[T], 0, len(s))
+// List returns the contents as a sorted T slice by given OrderedFuncs.
+func (s Set[T]) List(orderedFuncs OrderedFuncs[T]) []T {
```

Comment:
We can't change the signature of this function like this without breaking call sites.

Reply:
But this code is entirely new, isn't it?

Reply:
I thought the point was to delete the generate set types?

Reply:
OK Yeah, we have to update all instantiation and where the type isn't inferred, but we should be able to keep the method calls intact. I agree.

Reply:
Ideally we can delete the generated ones and use the type alias feature!

Reply:
@MikeSpreitzer @lavalamp
We can not sort any comparable elems when `List()` without giving less func. So how about changing the List behavior as that "returns the slice with contents in random order". 

In addition, a method `SortedList()` was introduce to sort slice by given less func. 

And `UnsortedList` should be deprecated.

```go
// List returns the slice with contents in random order.
func (s Set[T]) List() []T {
	res := make([]T, 0, len(s))
	for key := range s {
		res = append(res, key)
	}
	return res
}

// SortedList returns the contents as a sorted T slice by given less func.
func (s Set[T]) SortedList(less func(slice []T, i, j int) bool) []T {
	res := s.List()
	sort.Slice(res, func(i, j int) bool {
		return less(res, i, j)
	})
	return res
}

// UnsortedList returns the slice with contents in random order.
// Deprecated: use s.List() instead.
func (s Set[T]) UnsortedList() []T {
	return s.List()
}
```

Reply:
`ordered` lets us use the comparison operators. `comparable` means only equality comparison works. I didn't see in the other comment thread why we switched away from ordered?

If you want to let this work with non-ordered types I think the way to get an order function is to require it to be part of the interface, or otherwise extract it out of the type.

Otherwise this method signature has to change and I don't think we should do that. If we're going to do that, instead of making the caller tell us how to order, we should just change the name to `ArbitrarilyOrderedList` and not do this at all.